rate_limiter/src/group.rs

```diff
@@ -0,0 +1,531 @@
+/// used in exactly the same way as the RateLimiter type. When the RateLimiter
+/// within a RateLimiterGroup is unblocked, each RateLimiterGroupHandle will
+/// be notified.
+pub struct RateLimiterGroupHandle {
+    eventfd: Arc<EventFd>,
+    inner: Arc<RateLimiterGroupInner>,
+}
+
+impl RateLimiterGroupHandle {
+    fn new(inner: Arc<RateLimiterGroupInner>) -> result::Result<Self, Error> {
+        let eventfd = Arc::new(EventFd::new(0).map_err(Error::EventFd)?);
+        inner.add_eventfd(eventfd.clone());
+        Ok(Self { eventfd, inner })
+    }
+
+    /// Attempts to consume tokens and returns whether that is possible.
+    ///
+    /// If rate limiting is disabled on provided `token_type`, this function will always succeed.
+    pub fn consume(&self, tokens: u64, token_type: TokenType) -> bool {
+        self.inner.rate_limiter.consume(tokens, token_type)
+    }
+
+    /// Adds tokens of `token_type` to their respective bucket.
+    ///
+    /// Can be used to *manually* add tokens to a bucket. Useful for reverting a
+    /// `consume()` if needed.
+    pub fn manual_replenish(&self, tokens: u64, token_type: TokenType) {
+        self.inner.rate_limiter.manual_replenish(tokens, token_type)
+    }
+
+    /// This function needs to be called every time there is an event on the
+    /// FD provided by this object's `AsRawFd` trait implementation.
+    ///
+    /// # Errors
+    ///
+    /// If the rate limiter is disabled or is not blocked, an error is returned.
+    pub fn event_handler(&self) -> Result<(), Error> {
+        self.eventfd.read().map_err(Error::EventFdRead).map(|_| ())
+    }
+
+    /// Returns whether this rate limiter is blocked.
+    ///
+    /// The limiter 'blocks' when a `consume()` operation fails because there was not enough
+    /// budget for it.
+    /// An event will be generated on the exported FD when the limiter 'unblocks'.
+    pub fn is_blocked(&self) -> bool {
+        self.inner.rate_limiter.is_blocked()
+    }
+}
+
+impl Clone for RateLimiterGroupHandle {
+    fn clone(&self) -> Self {
+        RateLimiterGroupHandle::new(self.inner.clone()).unwrap()
+    }
+}
+
+impl AsRawFd for RateLimiterGroupHandle {
+    fn as_raw_fd(&self) -> std::os::unix::prelude::RawFd {
+        self.eventfd.as_raw_fd()
+    }
+}
+
+impl Drop for RateLimiterGroupHandle {
+    fn drop(&mut self) {
+        self.inner.remove_eventfd(self.eventfd.clone());
+    }
+}
+
+struct RateLimiterGroupInner {
+    rate_limiter: RateLimiter,
+    handles: Mutex<Vec<Arc<EventFd>>>,
+}
+
+impl RateLimiterGroupInner {
+    fn add_eventfd(&self, fd: Arc<EventFd>) {
+        let mut handles = self.handles.lock().unwrap();
+        handles.push(fd);
+    }
+
+    fn remove_eventfd(&self, fd: Arc<EventFd>) {
+        let mut handles = self.handles.lock().unwrap();
+        let index = handles
+            .iter()
+            .position(|handle| handle.as_raw_fd() == fd.as_raw_fd())
+            .unwrap();
+        handles.remove(index);
+    }
```

Comment:
Let's inline them to the constructor and destructor of `RateLimiterGroupHandle` directly which reads better to me.

Reply:
👍 . Inlined